File: Two Pointer/reversePrefix.js
const reversePrefix = (word, ch) => {
  const wordArr = word.split("");
  let left = 0;
  for (let right = 0; right < word.length; right++) {
    if (word[right] === ch) {
      while (left < right) {
        [wordArr[left], wordArr[right]] = [wordArr[right], wordArr[left]];
        left++;
        right--;
      }
      return wordArr.join("");
    }
  }
  return word;
};

// Time: O(n) where n is the length of the word
// Space: O(n) for the array to hold the characters
console.log(reversePrefix("abcdefd", "d")); // "dcbaefd"

File: Two Pointer/reverse_prefix.py
class ReversePrefix:
  def reversePrefix(self, word: str, ch: str) -> str:
    left = 0
    wordList = list(word)
    for right in range(len(word)):
      if word[right] == ch:
        while left < right:
          wordList[left], wordList[right] = wordList[right], wordList[left]
          left += 1
          right -= 1
        return "".join(wordList)
      
    return word

# Time complexity: O(n) where n is the length of the word.
# Space complexity: O(n) for the list conversion of the string.
word = "abcdefd"
ch = "d"

print(ReversePrefix().reversePrefix(word, ch))